### PR TITLE
fix: added unitttests

### PR DESCRIPTION
**Pull request by shnaaz**
Part of https://github.com/webex/webex-js-sdk/pull/2694
Added unittests for muteState.ts and webrtc-core.js

### CODE REVIEW

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js [hunk 1]
```diff
@@ -1,45 +1,130 @@
...
-          assert.equal(track.isUnmuteAllowed(), false);
-          await assert.isRejected(track.setMuted(false), Error, 'Unmute is not allowed');
-        });
+      it('rejects setMute(false) if unmute is not allowed', async () => {
+        track.setUnmuteAllowed(false);
 
-        it('resolves setMute(false) if unmute is allowed', async () => {
-          track.setUnmuteAllowed(true);
+        assert.equal(track.isUnmuteAllowed(), false);
+        const fn = () => track.setMuted(false);
```
Comment (marcin-bazyl) [inline packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:37]: why this change?

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js [hunk 2]
```diff
@@ -1,45 +1,130 @@
...
-          await track.setMuted(false);
-        })
+      it('resolves setMute(false) if unmute is allowed', async () => {
+        track.setUnmuteAllowed(true);
+
+        assert.equal(track.isUnmuteAllowed(), true);
+        track.setMuted(false);
+      });
+
+      it('tests setServerMuted- true to false', async () => {
```
Comment (marcin-bazyl) [inline packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:48]: all these setServerMuted tests could be just wrapped with a describe() and then no need to repeat "tests setServerMuted" in each test title

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js [hunk 3]
```diff
@@ -1,45 +1,130 @@
...
+        track.underlyingTrack.enabled = true; //start with unmuted
```
Comment (marcin-bazyl) [inline packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:49]: why do we need this? isn't the call to setMuted() enough to set the right state?

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js [hunk 4]
```diff
@@ -1,45 +1,130 @@
...
+
+        track.setMuted(true); //this.muted is true, call setServerMuted with false
+        assert.equal(track.muted, true);
+
+        const spy = sinon.stub(track, 'emit');
```
Comment (marcin-bazyl) [inline packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:54]: I think it would be better to just register an event handler for the event we are expecting and check that it gets called rather than stubbing the emit() method

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js [hunk 5]
```diff
@@ -1,45 +1,130 @@
...
+        track.setServerMuted(true, 'remotelyMuted');
+
+        assert.equal(track.muted, true);
+
+        const spyCall = spy.getCall(1);
+        assert.equal(spyCall, null);
+      });
+
+      it('tests setServerMuted- false to false', async () => {
```
Comment (marcin-bazyl) [inline packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:93]: this test and 3 previous ones are almost exactly the same, they could be refactored to re-use code from same helper test function

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 6]
```diff
@@ -509,3 +509,221 @@ describe('plugin-meetings', () => {
...
+
+  tests.forEach(({mediaType, title}) =>
+    describe(title, () => {
+
+      afterEach(() => {
+        MeetingUtil.remoteUpdateAudioVideo = originalRemoteUpdateAudioVideo;
+      });
+
+
+      it('tests handleLocalTrackChange', async () => {
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:583]: title should say "handleLocalTrackChange calls init()" or even better it should be:
```
describe("#handleLocalTrackChange", () => {
  it('calls init()', () => ....);
})
```

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 7]
```diff
@@ -509,3 +509,221 @@ describe('plugin-meetings', () => {
...
+    const direction = mediaType === AUDIO ? {sendAudio: true} : {sendVideo: true};
+    muteState = createMuteState(mediaType, meeting, direction, false);
+
+    originalRemoteUpdateAudioVideo = MeetingUtil.remoteUpdateAudioVideo;
+
+    MeetingUtil.remoteUpdateAudioVideo = sinon.stub().resolves(fakeLocus);
+  }
+
+   const setupSpies = (muteState, mediaType) => {
+    applyUnmuteAllowedToTrackSpy = sinon.spy(muteState, 'applyUnmuteAllowedToTrack');
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:562]: applyUnmuteAllowedToTrack, muteLocalTrack and applyClientStateToServer are all private functions. I don't think we should be spying on them - this makes the tests very closely coupled with the implementation, instead we should be mocking/spying the external methods/functions that MuteState calls, so for example instead of muteLocalTrack we should spy on the setServerMuted() method of the right track, instead of applyUnmuteAllowedToTrack we should spy on setUnmuteAllowed() on the right track, instead of applyClientStateToServer we should be checking that the right requests are sent to the server

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 8]
```diff
@@ -509,3 +509,221 @@ describe('plugin-meetings', () => {
...
+
+
+      it('tests handleLocalTrackChange', async () => {
+        setup(mediaType);
+        const spy = sinon.spy(muteState, 'init');
+        muteState.handleLocalTrackChange(meeting);
+        assert.calledOnceWithExactly(spy, meeting);
+      });
+
+      it('tests init when track is undefined', async () => {
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:590]: all tests that start with "tests init" should be wrapped with describe('#init')

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 9]
```diff
@@ -509,3 +509,221 @@ describe('plugin-meetings', () => {
...
+        muteState.init(meeting);
+
+        assert.calledOnceWithExactly(applyUnmuteAllowedToTrackSpy, meeting);
+        assert.notCalled(muteLocalTrackSpy);
+        assert.notCalled(setServerMutedSpy);
+        assert.calledOnceWithExactly(applyClientStateToServerSpy, meeting);
+        assert.isFalse(muteState.state.client.localMute);
+      });
+
+      it('#muteLocalTrack', async () => {
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:628]: muteLocalTrack is a private function, I don't think we should be testing it, but the test is actually calling init() instead, so I'm confused, what is being tested here? same for the next test

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 10]
```diff
@@ -509,3 +509,221 @@ describe('plugin-meetings', () => {
...
+          muteState.state.client.localMute = false;
+
+          const spy = sinon.spy(muteState, 'applyClientStateToServer');
+          muteState.handleLocalTrackMuteStateChange(meeting, true);
+          assert.equal(muteState.state.client.localMute, true);
+          assert.calledOnceWithExactly(spy, meeting)
+        });
+      });
+
+      describe('#applyClientStateLocally', () => {
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:692]: it's not clear to me what is the purpose of the tests inside this describe from the their titles. Are you sure you meant "ignoreMuteStateChange" in the titles and not sdkOwnsLocalTrack? Can you make them more descriptive?

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 11]
```diff
@@ -509,3 +509,221 @@ describe('plugin-meetings', () => {
...
+          sinon.restore();
+        });
+
+        it('checks when ignoreMuteStateChange is false', () => {
+          setup(mediaType);
+          setupSpies(muteState, mediaType);
+          muteState.sdkOwnsLocalTrack= false;
+
+          muteState.applyClientStateLocally(meeting, 'somereason');
+          assert.calledOnceWithExactly(muteLocalTrackSpy, meeting, muteState.state.client.localMute, 'somereason');
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:704]: we should be checking that setMuted was called on the track instead of a private muteLocalTrack

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js (hunk 9 above)
Comment (shnaaz) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:628]: I am trying to specifically test test muteLocalTrack by calling it through muteState.init(meeting) since it is a private method.

Comment (shnaaz) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:628]: If we do not want specific its here then I can add checks for serverMted in tests above

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js (hunk 1 above)
Comment (shnaaz) [reply packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:37]: Because the test was failing

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js (hunk 3 above)
Comment (shnaaz) [reply packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:49]: Here's  the reason
image

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js (hunk 1 above)
Comment (marcin-bazyl) [reply packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:37]: ah, yes, sorry I forgot to update it when I changed setMuted to not return a promise

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js [hunk 12]
```diff
@@ -1,45 +1,140 @@
-import {assert} from '@webex/test-helper-chai';
+import {assert, expect} from '@webex/test-helper-chai';
 import sinon from 'sinon';
-import {LocalCameraTrack, LocalMicrophoneTrack} from '@webex/media-helpers';
+import {LocalCameraTrack, LocalMicrophoneTrack, LocalMicrophoneTrackEvents, LocalCameraTrackEvents, LocalDisplayTrack, createCameraTrack, createMicrophoneTrack, createDisplayTrack} from '@webex/media-helpers';
+import * as wcmetracks from '@webex/internal-media-core';
 
-describe('media-helpers', () => {
+describe.only('media-helpers', () => {
```
Comment (marcin-bazyl) [inline packages/@webex/media-helpers/test/unit/spec/webrtc-core.js:6]: another .only

File: packages/@webex/media-helpers/test/unit/spec/webrtc-core.js (hunk 3 above)
Comment (shnaaz) [reply packages/@webex/media-helpers/test/unit/spec/webrtc-core.js]: setMuted calls super.setMuted which as you see in screenshot does something conditionally, hence setMuted isn't doing anything as this condition is not satisfied.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 13]
```diff
@@ -509,3 +509,202 @@ describe('plugin-meetings', () => {
...
+          sinon.restore();
+        });
+
+        it('checks when sdkOwnsLocalTrack is false', () => {
+          setup(mediaType);
+          setupSpies(muteState, mediaType);
+          muteState.sdkOwnsLocalTrack= false;
+
+          muteState.applyClientStateLocally(meeting, 'somereason');
+          assert.calledOnceWithExactly(setServerMutedSpy, muteState.state.client.localMute, 'somereason');
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:686]: it would be good to check that setMuted was not called here

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js [hunk 14]
```diff
@@ -509,3 +509,202 @@ describe('plugin-meetings', () => {
...
+  
+        it('tests when remoteMuted is true', async () => {
+          // testing that muteLocalTrack is called
+          setup(mediaType, true);
+          setupSpies(muteState, mediaType);
+  
+          muteState.init(meeting);
+  
+          const serverMutedSpyCall1 = setServerMutedSpy.getCall(0);
+          const serverMutedSpyCall2 = setServerMutedSpy.getCall(1);
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:633]: why are we expecting 2 calls to setServerMuted? the init() should trigger just 1 call to it.
I've checked this and it's because the first call is triggered from init() called from createMuteState() that's called by setup(), so I think we should call setServerMutedSpy.resetHistory() in setup() after calling createMuteState() (or inside setupSpies()) so that we have a clean slate for the test. We should probably do it for all the spies.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js (hunk 13 above)
Comment (shnaaz) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:686]: done

**Review by marcin-bazyl — APPROVED** [review]
